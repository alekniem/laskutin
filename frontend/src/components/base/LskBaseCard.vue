<template>
    <v-card v-bind="$attrs">
        <slot name="header" />
        <v-divider />
        <slot />
    </v-card>
</template>

<script>
export default {
    name: "LskBaseCard",
};
</script>
