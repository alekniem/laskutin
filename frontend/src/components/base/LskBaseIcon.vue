<template>
    <v-tooltip bottom>
        <template #activator="{ attrs, on }">
            <v-icon
                @click="clickFunction(parameter)"
                :class="cssClass"
                small
                v-bind="attrs"
                v-on="on"
            >
                {{ iconName }}
            </v-icon>
        </template>
        <span>{{ title | trans }}</span>
    </v-tooltip>
</template>

<script>
export default {
    name: "LskBaseIcon",

    props: {
        clickFunction: { type: Function, required: true },
        cssClass: { type: String, required: false },
        iconName: { type: String, required: true },
        parameter: { type: Object, required: true },
        title: { type: String, required: true },
    },
};
</script>
