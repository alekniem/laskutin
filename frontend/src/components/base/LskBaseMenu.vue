<template>
    <v-menu bottom left>
        <template #activator="{ attrs, on }">
            <v-btn :id="activatorId" icon v-bind="attrs" v-on="on">
                <v-icon>mdi-dots-vertical</v-icon>
            </v-btn>
        </template>
        <v-list>
            <slot />
        </v-list>
    </v-menu>
</template>

<script>
export default {
    name: "LskBaseMenu",

    props: {
        activatorId: { type: String, required: true },
    },
};
</script>
