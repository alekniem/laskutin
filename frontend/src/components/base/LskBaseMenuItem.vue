<template>
    <v-list-item :id="cid" link>
        <v-list-item-title>
            {{ text | trans }}
        </v-list-item-title>
    </v-list-item>
</template>

<script>
export default {
    name: "LskBaseMenuItem",

    props: {
        cid: { type: String, required: true },
        text: { type: String, required: true },
    },
};
</script>
