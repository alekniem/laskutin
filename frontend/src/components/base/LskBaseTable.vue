<template>
    <v-data-table
        :headers="headers"
        :id="cid"
        :items="items"
        :items-per-page="10"
        dense
        sort-by="id"
        sort-desc
        v-bind="$attrs"
    >
        <template #[`body.append`]>
            <slot name="bodyAppend" />
        </template>
        <template #[`item.actions`]="{ item }">
            <LskBaseIcon
                v-for="(action, key) in actions"
                :clickFunction="action.clickFunction"
                :cssClass="key + 1 < actions.length ? 'mr-3' : ''"
                :iconName="action.iconName"
                :key="key"
                :parameter="item"
                :title="action.title"
            />
        </template>
        <template #[`item.amount`]="{ item }">
            {{ item.amount | centsToEuros }}
        </template>
        <template #[`item.due_date`]="{ item }">
            {{ item.due_date | formatDate("D.M.YYYY") }}
        </template>
        <template #[`item.invoice_date`]="{ item }">
            {{ item.invoice_date | formatDate("D.M.YYYY") }}
        </template>
        <template #[`item.total_amount`]="{ item }">
            {{ item.total_amount | centsToEuros }}
        </template>
        <template #[`item.total_amount_sum`]="{ item }">
            {{ item.total_amount_sum | centsToEuros }}
        </template>
        <template #[`item.vat_amount`]="{ item }">
            {{ item.vat_amount | centsToEuros }}
        </template>
        <template #no-data>
            <LskBaseError v-if="errorMessage">
                {{ errorMessage | trans }}
            </LskBaseError>
        </template>
    </v-data-table>
</template>

<script>
export default {
    name: "LskBaseTable",

    props: {
        actions: { type: Array, required: false, default: () => [] },
        cid: { type: String, required: true },
        errorMessage: { type: String, required: false },
        headings: { type: Object, required: true },
        tableData: { type: Array, required: true },
    },

    computed: {
        headers() {
            let result = [];
            Object.keys(this.headings).forEach((key) => {
                result.push({
                    class: "blue--text text--darken-2",
                    text: this.trans(this.headings[key]),
                    value: key,
                });
            });
            if (this.actions.length) {
                result.push({
                    class: "blue--text text--darken-2",
                    sortable: false,
                    text: this.trans("Actions"),
                    value: "actions",
                });
            }
            return result;
        },

        items() {
            return this.tableData.map((data) => {
                let result = Object.assign({}, data);
                if (this.actions.length) {
                    result.actions = "";
                }
                return result;
            });
        },
    },
};
</script>
