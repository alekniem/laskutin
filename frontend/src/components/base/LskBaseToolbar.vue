<template>
    <v-toolbar flat>
        <v-toolbar-title :id="titleId">
            <slot />
        </v-toolbar-title>
        <v-spacer />
        <slot name="button" />
    </v-toolbar>
</template>

<script>
export default {
    name: "LskBaseToolbar",

    props: {
        titleId: { type: String, required: true },
    },
};
</script>
