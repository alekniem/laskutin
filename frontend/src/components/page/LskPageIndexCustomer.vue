<template>
    <v-row>
        <v-col cols="12">
            <LskBaseCard>
                <template #header>
                    <LskBaseToolbar :titleId="cid + 'Title'">
                        <template #button>
                            <LskBaseMenu :activatorId="cid + 'MenuActivator'">
                                <LskBaseMenuItem
                                    @click.native="addItem"
                                    :cid="cid + 'AddCustomerMenuItem'"
                                    text="Add new customer"
                                />
                            </LskBaseMenu>
                        </template>
                        {{ "Customers" | trans }}
                    </LskBaseToolbar>
                </template>
                <LskBaseTable
                    :actions="actions"
                    :cid="cid + 'Table'"
                    :errorMessage="errorMessage"
                    :headings="headings"
                    :tableData="tableData"
                />
                <LskDialogAddOrEditCustomer
                    :cid="cid + 'AddCustomerDialog'"
                    :key="keyAddDialog"
                    v-model="showAddDialog"
                />
                <LskDialogDelete
                    :cid="cid + 'DeleteCustomerDialog'"
                    :id="selectedId"
                    :key="keyDeleteDialog"
                    apiActionName="customerDestroy"
                    checkboxLabel="Yes, delete customer"
                    title="Delete customer"
                    v-model="showDeleteDialog"
                />
            </LskBaseCard>
        </v-col>
    </v-row>
</template>

<script>
import LskIndex from "@/mixins/LskIndex";

export default {
    name: "LskPageIndexCustomer",

    mixins: [LskIndex],

    data() {
        return {
            apiActionName: "customerIndex",
            cid: "customerIndexPage",
            headings: {
                id: "Customer id",
                name: "Name",
                address_line_one: "Address line 1",
                address_line_two: "Address line 2",
            },
            showItemRouteName: "showCustomer",
        };
    },
};
</script>
