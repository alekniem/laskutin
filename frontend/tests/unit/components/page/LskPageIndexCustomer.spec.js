import axios from "axios";
import flushPromises from "flush-promises";
import sinon from "sinon";
import LskBaseCard from "@/components/base/LskBaseCard";
import LskBaseIcon from "@/components/base/LskBaseIcon";
import LskBaseMenu from "@/components/base/LskBaseMenu";
import LskBaseMenuItem from "@/components/base/LskBaseMenuItem";
import LskBaseTable from "@/components/base/LskBaseTable";
import LskBaseToolbar from "@/components/base/LskBaseToolbar";
import LskPageIndexCustomer from "@/components/page/LskPageIndexCustomer";
import { createWrapper } from "UT/lsk_test_helper";

jest.mock("axios");

axios.get.mockImplementation((url) => {
    if (url === "customer") {
        return Promise.resolve({
            data: [
                {
                    id: 1,
                    name: "NAME 1",
                    address_line_one: "ADDRESS LINE ONE 1",
                    address_line_two: "ADDRESS LINE TWO 1",
                },
                {
                    id: 2,
                    name: "NAME 2",
                    address_line_one: "ADDRESS LINE ONE 2",
                    address_line_two: "ADDRESS LINE TWO 2",
                },
            ],
        });
    }
    return Promise.reject({
        response: {
            data: {
                message: "GET QUERY ERROR",
            },
        },
    });
});

const stubs = {
    LskBaseCard,
    LskBaseIcon,
    LskBaseMenu,
    LskBaseMenuItem,
    LskBaseTable,
    LskBaseToolbar,
    LskDialogAddOrEditCustomer: true,
    LskDialogDelete: true,
};

describe("LskPageIndexCustomer", () => {
    afterEach(() => {
        sinon.restore();
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskPageIndexCustomer, {
            stubs,
        });
        await flushPromises();
        expect(wrapper.html()).toMatchSnapshot("page");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskPageIndexCustomer, {
            stubs,
        });
        await flushPromises();

        const menu = wrapper.findComponent(LskBaseMenu);
        menu.find("#customerIndexPageMenuActivator").trigger("click");
        await flushPromises();
        expect(menu.html()).toMatchSnapshot("menu");

        menu.find("#customerIndexPageAddCustomerMenuItem").trigger("click");
        await flushPromises();
        const dialog = wrapper.find("LskDialogAddOrEditCustomer-stub");
        expect(dialog.html()).toMatchSnapshot("dialog add customer");
    });

    it("should match snapshot", async () => {
        const wrapper = createWrapper(LskPageIndexCustomer, {
            stubs,
        });
        await flushPromises();
        const table = wrapper.findComponent(LskBaseTable);
        table.find(".mdi-delete").trigger("click");
        await flushPromises();
        const dialog = wrapper.find("LskDialogDelete-stub");
        expect(dialog.html()).toMatchSnapshot("dialog delete customer");
    });

    it("should call router", async () => {
        const wrapper = createWrapper(LskPageIndexCustomer, {
            stubs,
        });
        await flushPromises();
        wrapper.vm.$router = { push: sinon.spy() };
        const table = wrapper.findComponent(LskBaseTable);
        table.find(".mdi-monitor").trigger("click");
        await flushPromises();
        const param = {
            name: "showCustomer",
            params: { id: 2 },
        };
        expect(wrapper.vm.$router.push.calledWith(param)).toBe(true);
    });
});
