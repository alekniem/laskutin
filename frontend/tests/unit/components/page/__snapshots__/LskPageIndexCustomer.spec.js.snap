// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LskPageIndexCustomer should match snapshot: dialog add customer 1`] = `
<lskdialogaddoreditcustomer-stub cid="customerIndexPageAddCustomerDialog"
                                 value="true"
>
</lskdialogaddoreditcustomer-stub>
`;

exports[`LskPageIndexCustomer should match snapshot: dialog delete customer 1`] = `
<lskdialogdelete-stub cid="customerIndexPageDeleteCustomerDialog"
                      id="2"
                      apiactionname="customerDestroy"
                      checkboxlabel="Yes, delete customer"
                      title="Delete customer"
                      value="true"
>
</lskdialogdelete-stub>
`;

exports[`LskPageIndexCustomer should match snapshot: menu 1`] = `
<div class="v-menu">
  <transition-stub name="v-menu-transition">
    <div role="menu"
         class="v-menu__content theme--light menuable__content__active "
         style="max-height: auto; min-width: 0px; max-width: auto; top: 12px; left: 12px; transform-origin: top left; z-index: 8; display: none;"
    >
      <div class="v-list v-sheet theme--light">
        <div tabindex="0"
             id="customerIndexPageAddCustomerMenuItem"
             role="menuitem"
             class="v-list-item v-list-item--link theme--light"
        >
          <div class="v-list-item__title">
            Add new customer
          </div>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`LskPageIndexCustomer should match snapshot: page 1`] = `
<div class="row">
  <div class="col col-12">
    <div class="v-card v-sheet theme--light">
      <header class="v-sheet theme--light v-toolbar v-toolbar--flat"
              style="height: 64px;"
      >
        <div class="v-toolbar__content"
             style="height: 64px;"
        >
          <div id="customerIndexPageTitle"
               class="v-toolbar__title"
          >
            Customers
          </div>
          <div class="spacer">
          </div>
          <div class="v-menu">
          </div>
          <button type="button"
                  class="v-btn v-btn--icon v-btn--round theme--light v-size--default"
                  id="customerIndexPageMenuActivator"
                  role="button"
                  aria-haspopup="true"
                  aria-expanded="false"
          >
            <span class="v-btn__content">
              <i aria-hidden="true"
                 class="v-icon notranslate mdi mdi-dots-vertical theme--light"
              >
              </i>
            </span>
          </button>
        </div>
      </header>
      <hr role="separator"
          aria-orientation="horizontal"
          class="v-divider theme--light"
      >
      <div class="v-data-table v-data-table--dense v-data-table--has-bottom theme--light"
           id="customerIndexPageTable"
      >
        <div class="v-data-table__wrapper">
          <table>
            <colgroup>
              <col class>
              <col class>
              <col class>
              <col class>
              <col class>
            </colgroup>
            <thead class="v-data-table-header">
              <tr>
                <th role="columnheader"
                    scope="col"
                    aria-label="Customer id: Sorted descending. Activate to remove sorting."
                    aria-sort="descending"
                    class="text-start blue--text text--darken-2 sortable active desc"
                >
                  <span>
                    Customer id
                  </span>
                  <i aria-hidden="true"
                     class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                     style="font-size: 18px;"
                  >
                  </i>
                </th>
                <th role="columnheader"
                    scope="col"
                    aria-label="Name: Not sorted. Activate to sort ascending."
                    aria-sort="none"
                    class="text-start blue--text text--darken-2 sortable"
                >
                  <span>
                    Name
                  </span>
                  <i aria-hidden="true"
                     class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                     style="font-size: 18px;"
                  >
                  </i>
                </th>
                <th role="columnheader"
                    scope="col"
                    aria-label="Address line 1: Not sorted. Activate to sort ascending."
                    aria-sort="none"
                    class="text-start blue--text text--darken-2 sortable"
                >
                  <span>
                    Address line 1
                  </span>
                  <i aria-hidden="true"
                     class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                     style="font-size: 18px;"
                  >
                  </i>
                </th>
                <th role="columnheader"
                    scope="col"
                    aria-label="Address line 2: Not sorted. Activate to sort ascending."
                    aria-sort="none"
                    class="text-start blue--text text--darken-2 sortable"
                >
                  <span>
                    Address line 2
                  </span>
                  <i aria-hidden="true"
                     class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light"
                     style="font-size: 18px;"
                  >
                  </i>
                </th>
                <th role="columnheader"
                    scope="col"
                    aria-label="Actions"
                    class="text-start blue--text text--darken-2"
                >
                  <span>
                    Actions
                  </span>
                </th>
              </tr>
            </thead>
            <tbody>
              <tr class>
                <td class="text-start">
                  2
                </td>
                <td class="text-start">
                  NAME 2
                </td>
                <td class="text-start">
                  ADDRESS LINE ONE 2
                </td>
                <td class="text-start">
                  ADDRESS LINE TWO 2
                </td>
                <td class="text-start">
                  <button type="button"
                          aria-haspopup="true"
                          aria-expanded="false"
                          class="v-icon notranslate v-icon--link mdi mdi-monitor theme--light mr-3"
                          style="font-size: 16px;"
                  >
                  </button>
                  <span class="v-tooltip v-tooltip--bottom">
                  </span>
                  <span class="v-tooltip v-tooltip--bottom">
                  </span>
                  <button type="button"
                          aria-haspopup="true"
                          aria-expanded="false"
                          class="v-icon notranslate v-icon--link mdi mdi-delete theme--light"
                          style="font-size: 16px;"
                  >
                  </button>
                </td>
              </tr>
              <tr class>
                <td class="text-start">
                  1
                </td>
                <td class="text-start">
                  NAME 1
                </td>
                <td class="text-start">
                  ADDRESS LINE ONE 1
                </td>
                <td class="text-start">
                  ADDRESS LINE TWO 1
                </td>
                <td class="text-start">
                  <button type="button"
                          aria-haspopup="true"
                          aria-expanded="false"
                          class="v-icon notranslate v-icon--link mdi mdi-monitor theme--light mr-3"
                          style="font-size: 16px;"
                  >
                  </button>
                  <span class="v-tooltip v-tooltip--bottom">
                  </span>
                  <span class="v-tooltip v-tooltip--bottom">
                  </span>
                  <button type="button"
                          aria-haspopup="true"
                          aria-expanded="false"
                          class="v-icon notranslate v-icon--link mdi mdi-delete theme--light"
                          style="font-size: 16px;"
                  >
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div class="v-data-footer">
          <div class="v-data-footer__select">
            Rows per page:
            <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
              <div class="v-input__control">
                <div role="button"
                     aria-haspopup="listbox"
                     aria-expanded="false"
                     aria-owns="list-22"
                     class="v-input__slot"
                >
                  <div class="v-select__slot">
                    <div class="v-select__selections">
                      <div class="v-select__selection v-select__selection--comma">
                        10
                      </div>
                      <input aria-label="Rows per page:"
                             id="input-22"
                             readonly="readonly"
                             type="text"
                             aria-readonly="false"
                             autocomplete="off"
                      >
                    </div>
                    <div class="v-input__append-inner">
                      <div class="v-input__icon v-input__icon--append">
                        <i aria-hidden="true"
                           class="v-icon notranslate mdi mdi-menu-down theme--light"
                        >
                        </i>
                      </div>
                    </div>
                    <input type="hidden"
                           value="10"
                    >
                  </div>
                  <div class="v-menu">
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-data-footer__pagination">
            1-2 of 2
          </div>
          <div class="v-data-footer__icons-before">
            <button type="button"
                    disabled="disabled"
                    class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
                    aria-label="Previous page"
            >
              <span class="v-btn__content">
                <i aria-hidden="true"
                   class="v-icon notranslate mdi mdi-chevron-left theme--light"
                >
                </i>
              </span>
            </button>
          </div>
          <div class="v-data-footer__icons-after">
            <button type="button"
                    disabled="disabled"
                    class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
                    aria-label="Next page"
            >
              <span class="v-btn__content">
                <i aria-hidden="true"
                   class="v-icon notranslate mdi mdi-chevron-right theme--light"
                >
                </i>
              </span>
            </button>
          </div>
        </div>
      </div>
      <lskdialogaddoreditcustomer-stub cid="customerIndexPageAddCustomerDialog">
      </lskdialogaddoreditcustomer-stub>
      <lskdialogdelete-stub cid="customerIndexPageDeleteCustomerDialog"
                            apiactionname="customerDestroy"
                            checkboxlabel="Yes, delete customer"
                            title="Delete customer"
      >
      </lskdialogdelete-stub>
    </div>
  </div>
</div>
`;
